Nodejs/Product/Nodejs/Workspace/PackageJsonTestContainerDiscoverer.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestWindow.Extensibility;
+using Microsoft.VisualStudio.Workspace;
+using Microsoft.VisualStudio.Workspace.Indexing;
+using Microsoft.VisualStudio.Workspace.VSIntegration.Contracts;
+
+namespace Microsoft.NodejsTools.Workspace
+{
+    [Export(typeof(ITestContainerDiscoverer))]
+    public sealed class PackageJsonTestContainerDiscoverer : ITestContainerDiscoverer
+    {
+        private readonly IVsFolderWorkspaceService workspaceService;
+
+        private readonly List<PackageJsonTestContainer> containers = new List<PackageJsonTestContainer>();
+        private readonly object containerLock = new object();
```

Comment:
>containerLock [](start = 32, length = 13)

I'm probably forgetting some framework guideline, but why not lock on `containers`?

Reply:
In this case locking on container would be fine, since it's not a public property. But I prefer to use the well-known pattern to avoid confusion in the future.

Reply:
I understood that pattern was for value types or public members of reference types, but I don't feel strongly about it.

---
In reply to: [192169424](https://github.com/Microsoft/nodejstools/pull/1959#discussion_r192169424) [](ancestors = 192169424)